### PR TITLE
Limited "Multiple plugin in same folder" support.

### PR DESCRIPTION
**Pull request by BaronGreenback**
Workaround for https://github.com/jellyfin/jellyfin/issues/4908

This permits multiple plugins in the same folder being loaded and accessed from the dashboard, with the following provisos:-

- A depreciated warning message will be displayed.
- The ability to disable / enable and delete these plugins is disabled, as no disk manifest can be created.
- Upgrading of these plugins via the web will not work as will probably cause "Bad Things"(TM)
- Plugins that fail can no longer be disabled, meaning that they will continue to be loaded every time JF restarts.

### CODE REVIEW

File: Emby.Server.Implementations/Plugins/PluginManager.cs [hunk 1]
```diff
@@ -105,18 +108,52 @@ public IEnumerable<Assembly> LoadAssemblies()
...
                     {
                         assembly = Assembly.LoadFrom(file);
 
                         // This force loads all reference dll's that the plugin uses in the try..catch block.
                         // Removing this will cause JF to bomb out if referenced dll's cause issues.
-                        assembly.GetExportedTypes();
+                        var ass = assembly.GetExportedTypes();
+
+                        // Backwards compatibility fix where we multiple assemblies containing types.
+                        var containsPlugin = ass.Any(i => pluginType.IsAssignableFrom(i));
```
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:126]: `i` is already in use here

Comment (BaronGreenback) [reply Emby.Server.Implementations/Plugins/PluginManager.cs:126]: change var name

File: Emby.Server.Implementations/Plugins/PluginManager.cs [hunk 2]
```diff
@@ -127,6 +164,13 @@ public IEnumerable<Assembly> LoadAssemblies()
                     yield return assembly;
                 }
             }
+
+            if (dynamicallyAdded > 0)
+            {
+                _logger.LogWarning("Multiple plugins sharing the same folder is depreciated.");
+                _logger.LogWarning("Running plugins in this configuration is no longer recommended, and may have unforeseen consequences.");
+                _logger.LogWarning("Please move each plugin listed above, along with any of their supporting assemblies, into their own folder, and restart Jellyfin.");
```
[the remark this answers is not in the document]
Comment (BaronGreenback) [reply Emby.Server.Implementations/Plugins/PluginManager.cs:172]: I'd originally put it as one with \r\n but then it didn't adhere to the normal log formatting. Remembering the plugin PR to reformat the logs, i opted for this.

File: Emby.Server.Implementations/Plugins/PluginManager.cs [hunk 3]
```diff
@@ -374,7 +418,7 @@ private bool ChangePluginState(LocalPlugin plugin, PluginStatus state)
         private LocalPlugin? GetPluginByAssembly(Assembly assembly)
         {
             // Find which plugin it is by the path.
-            return _plugins.FirstOrDefault(p => string.Equals(p.Path, Path.GetDirectoryName(assembly.Location), StringComparison.Ordinal));
+            return _plugins.FirstOrDefault(p => p.DllFiles.Contains(assembly.Location, StringComparer.Ordinal)));
```
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:421]:
```suggestion
            return _plugins.FirstOrDefault(p => p.DllFiles.Contains(assembly.Location, StringComparer.Ordinal));
```

Accidentally suggested extra `)`

File: Emby.Server.Implementations/Plugins/PluginManager.cs [hunk 4]
```diff
@@ -94,9 +93,13 @@ public IEnumerable<Assembly> LoadAssemblies()
                 }
             }
 
+            var pluginType = typeof(IPlugin);
+            int dynamicallyAdded = 0;
+
             // Now load the assemblies..
-            foreach (var plugin in _plugins)
+            for (int i = 0; i <= _plugins.Count - dynamicallyAdded - 1; i++)
```
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:100]:
```suggestion
            var originalPluginCount = _plugins.Count;
            for (int i = 0; i < originalPluginCount; i++)
```

And remove `dynamicallyAdded` counter

File: Emby.Server.Implementations/Plugins/PluginManager.cs [hunk 5]
```diff
@@ -105,18 +108,52 @@ public IEnumerable<Assembly> LoadAssemblies()
...
                     try
                     {
                         assembly = Assembly.LoadFrom(file);
 
                         // This force loads all reference dll's that the plugin uses in the try..catch block.
                         // Removing this will cause JF to bomb out if referenced dll's cause issues.
-                        assembly.GetExportedTypes();
+                        var ass = assembly.GetExportedTypes();
+
+                        // Backwards compatibility fix where we multiple assemblies containing types.
```
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:125]:
```suggestion
                        // Backwards compatibility fix where we have multiple assemblies containing types.
```

File: Emby.Server.Implementations/Plugins/PluginManager.cs [hunk 6]
```diff
@@ -105,18 +108,52 @@ public IEnumerable<Assembly> LoadAssemblies()
...
+                        var containsPlugin = ass.Any(a => pluginType.IsAssignableFrom(a));
+
+                        if (containsPlugin && ++pluginCount > 1)
+                        {
+                            _logger.LogWarning("Depreciated: Multiple plugins found in one folder : {Plugin}", file);
```
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:130]:
```suggestion
                            _logger.LogWarning("Depreciated: Multiple plugins found in one folder: {Plugin}", file);
```

File: Emby.Server.Implementations/Plugins/PluginManager.cs [hunk 7]
```diff
@@ -105,18 +108,52 @@ public IEnumerable<Assembly> LoadAssemblies()
...
+                            // Clone LocalPlugin so that it can be managed.
+                            var p = new LocalPlugin(
+                                plugin.Path,
+                                true,
+                                new PluginManifest()
+                                {
+                                    Status = PluginStatus.Active,
+                                    Name = Path.GetFileNameWithoutExtension(file) ?? file,
+                                    AutoUpdate = false,
+                                    Id = file.GetMD5(),
```
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:142]: The file likely already has an ID, so we should probably use it?

File: Emby.Server.Implementations/Plugins/PluginManager.cs (hunk 4 above)
Comment (BaronGreenback) [reply Emby.Server.Implementations/Plugins/PluginManager.cs:100]: done.

File: Emby.Server.Implementations/Plugins/PluginManager.cs (hunk 7 above)
Comment (BaronGreenback) [reply Emby.Server.Implementations/Plugins/PluginManager.cs:142]: ??
The original plugin has an ID which needs to be unique.
_file_ contains the string filename of the assembly that has been loaded. LocalPlugin.ID points to Manifest.ID, so it does appear anywhere else.

**Review by crobibero — APPROVED** [review]

File: MediaBrowser.Common/Plugins/PluginManifest.cs [hunk 8]
```diff
@@ -106,5 +106,12 @@ public PluginManifest()
         /// </summary>
         [JsonPropertyName("imagePath")]
         public string? ImagePath { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether this manifest was created from to the depreciated method of multiple plugins per folder.
+        /// In this instance manifests cannot be stored on disk.
+        /// </summary>
+        [JsonIgnore]
+        public bool Depreciated { get; set; }
```
Comment (thornbill) [inline MediaBrowser.Common/Plugins/PluginManifest.cs:115]: Should this be “Deprecated”?

Comment (BaronGreenback) [reply MediaBrowser.Common/Plugins/PluginManifest.cs:115]: Thanks - Spelling was never my forte!

**Review by crobibero — APPROVED** [review]

File: MediaBrowser.Common/Plugins/PluginManifest.cs (hunk 8 above)
Comment (cvium) [reply MediaBrowser.Common/Plugins/PluginManifest.cs:115]: What about the rest of the occurrences?

Comment (BaronGreenback) [reply MediaBrowser.Common/Plugins/PluginManifest.cs:115]: Each occurrence of a dll that contains an IPlugin which are found in the same folder, gets an in memory **only** instance of plugin manifest. The depreciated flag stops these from being written to disk - as they would overwrite each other.

Plugins from elsewhere operate as normal.

Comment (cvium) [reply MediaBrowser.Common/Plugins/PluginManifest.cs:115]: The spelling error 😉

Comment (BaronGreenback) [reply MediaBrowser.Common/Plugins/PluginManifest.cs]: fixed

**Review by barronpm — APPROVED** [review]

**Review by cvium — CHANGES_REQUESTED** [review]
I don't agree with keeping backwards compatibility for something as mundane as plugin folder structure. We can handle the migration automatically or ask people to fix it themselves.

**Comment by BaronGreenback** [comment]
Had a look at the possibility of implementing an auto migration option, and it's complex to say the least.

First of all, each dll has to be loaded into a reflection namespace so that it can be examines.
Next, if it's a plugin, then each of its assembly dependencies have to be examined to see if it had corresponding dll's that need to be moved with it. (Each dependency is a string - without path)
Third a folder structure has to be created, for which the version can only be done by parsing string (no property available)...
Then there's the file moving, and finally the unloading of the reflection namespace.
If dll's are shared between plugins, it will fail.
If the unloading has completed before the file move (done asynch), it will fail.
If the plugin type name doesn't match the plugin - it will put it into a weird folder.

With all this in mind, I suggest we rule the auto migration option.

**Pull request closed without merging**